HARK/agent_population.py

```diff
@@ -0,0 +1,116 @@
+from HARK.core import AgentType
+
+Parameters = NewType("Parameters", "dict")
+
+
+class AgentPopulation:
+    def __init__(
+        self,
+        agent_class: AgentType,
+        parameter_dict: Parameters,
+        t_cycle: int,
+        agent_class_count: int,
+    ):
+        self.agent_class = agent_class
+        self.parameter_dict = parameter_dict
+        self.t_cycle = t_cycle
+        self.agent_class_count = agent_class_count
+
+        self.time_var = agent_class.time_vary_
+        self.time_inv = agent_class.time_inv_
+
+    def parse_params(self):
+
+        param_dict = self.parameter_dict
+
+        agent_dicts = []
+        for agent in range(self.agent_class_count):
+            agent_params = {}
+
+            for key_param in param_dict:
+
+                if key_param in self.time_var:
+                    # parameters that vary over time have to be repeated
+                    parameter_per_t = []
+                    for t in range(self.t_cycle):
+                        if param_dict[key_param].kind == "agent":
+                            if param_dict[key_param].value is list:
+                                # if the parameter is a list, it's agent and time
+                                parameter_per_t.append(
+                                    param_dict[key_param].value[agent][t]
+                                )
+                            else:
+                                parameter_per_t.append(
+                                    param_dict[key_param].value[agent]
+                                )
+                        elif param_dict[key_param].kind == "time":
+                            # if kind is time, it applies to all agents but varies over time
+                            # assert param_dict[key_param].value is list
+                            parameter_per_t.append(param_dict[key_param].value[t])
+                        elif param_dict[key_param].kind == "fixed":
+                            # if kind is fixed, it applies to all agents at all times
+                            # assert param_dict[key_param].value is not list
+                            parameter_per_t.append(param_dict[key_param].value)
+
+                    agent_params[key_param] = parameter_per_t
+
+                elif key_param in self.time_inv:
+                    if param_dict[key_param].kind == "agent":
+                        # assert param_dict[key_param].value is list
+                        agent_params[key_param] = param_dict[key_param].value[agent]
+                    if param_dict[key_param].kind == "fixed":
+                        # assert param_dict[key_param].value is not list
+                        agent_params[key_param] = param_dict[key_param].value
+
+            agent_dicts.append(agent_params)
+
+        self.agent_dicts = agent_dicts
+
+    def create_distributed_agents(self):
+
+        self.agents = [self.agent_class(parameters) for parameters in self.agent_dicts]
```

Comment:
One thing that the current AgentPopulation does which I don't see in this code yet is manage the AgentCount for the total population and for the 'distributed' agents.

https://github.com/sbenthall/SHARKFin/blob/master/HARK/hark_portfolio_agents.py#L218-L236

I believe that for the `agent` parameters, if the value is a Distribution, then that should reflect the proportion of the population that has that parameter. So, for example, it should be in principle possible to have a discretized normal distribution of CRRA.

Reply:
I think the first one can be resolved with something like 

`parameters["AgentCount"] = hark_param([1,1,1], "agent")`

meaning there is 1 agent per distint agent type.

Reply:
I see how that would work.
However, it seems like it would be cumbersome to scale,

I.e., it would be nice to be able to start with a small, correctly proportioned population, and then once the bugs have been worked out, raise the TotalAgentCount and have everything just work.

Reply:
Along these lines ... with your first implementation, the modeler has to by hand guarantee that each `agent` parameter has the same number of entries in it (or the right number in the `agent` dimension).

Whereas it would be smoother, I think, if the modeler provided Distributions for the parameters, and then provided just one parameter for the number of bins to subdivide the population into.

Reply:
It's probably worth supporting both ways of doing it, really.